crates/ide/src/runnables.rs

```diff
@@ -575,6 +616,102 @@ fn has_test_function_or_multiple_test_submodules(
     number_of_test_submodules > 1
 }
 
+struct TestDefs<'a, 'b>(&'a Semantics<'b, RootDatabase>, hir::Crate, hir::FileRange);
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct UpdateTest {
+    pub expect_test: bool,
+    pub insta: bool,
+    pub snapbox: bool,
+}
+
+impl UpdateTest {
+    pub fn label(&self) -> Option<SmolStr> {
+        let mut builder: SmallVec<[_; 3]> = SmallVec::new();
+        if self.expect_test {
+            builder.push("Expect");
+        }
+        if self.insta {
+            builder.push("Insta");
+        }
+        if self.snapbox {
+            builder.push("Snapbox");
+        }
+
+        let res: SmolStr = builder.join(" + ").into();
+        res.is_empty().not().then_some(res)
+    }
+}
+
+impl<'a, 'b> TestDefs<'a, 'b> {
+    fn new(
+        sema: &'a Semantics<'b, RootDatabase>,
+        current_krate: hir::Crate,
+        file_range: hir::FileRange,
+    ) -> Self {
+        Self(sema, current_krate, file_range)
+    }
+
+    fn update_test(&self) -> UpdateTest {
+        UpdateTest { expect_test: self.expect_test(), insta: self.insta(), snapbox: self.snapbox() }
+    }
+
+    fn expect_test(&self) -> bool {
+        self.find_macro("expect_test", &["expect", "expect_file"])
+    }
+
+    fn insta(&self) -> bool {
+        self.find_macro(
+            "insta",
+            &[
+                "assert_snapshot",
+                "assert_debug_snapshot",
+                "assert_display_snapshot",
+                "assert_json_snapshot",
+                "assert_yaml_snapshot",
+                "assert_ron_snapshot",
+                "assert_toml_snapshot",
+                "assert_csv_snapshot",
+                "assert_compact_json_snapshot",
+                "assert_compact_debug_snapshot",
+                "assert_binary_snapshot",
+            ],
+        )
+    }
+
+    fn snapbox(&self) -> bool {
+        self.find_macro("snapbox", &["assert_data_eq", "file", "str"])
+    }
+
+    fn find_macro(&self, crate_name: &str, paths: &[&str]) -> bool {
```

Comment:
This seems a bit weird to me the way this works, a better idea would be to just check the macro calls of the corresponding body I think and match them up with the expected macros. That is fetch the `BodySourceMap` for the corresponding body (we might have that via semantics already here, not sure), and iterate the `expansions` in there to find the macros. Before doing so we should probably first try to resolve all the known testing macros and cache them to just match up the IDs against another

Reply:
I use FindUsage because it allows us to quickly filter out irrelevant macro calls by name. Since we need to search within (potentially large) mod/impl blocks, this approach is more efficient.

Reply:
> Resolving the testing macros and caching their IDs is reasonable, but I haven’t figured out where to store them yet

The caching here is good enough to be done at the start of the runnables computation (that is we don't cache across requests)

> I use FindUsage because it allows us to quickly filter out irrelevant macro calls by name.

I see, I was thinking of what if the macro call is hidden behind another macro call, but I think that basicalyl breaks all of those frameworks anyways so little point in trrying to do that. I see your approach then. The way this is done here still feels a bit brittle to me though. Will add another comment regarding that in a bit